src/app/util/ContentAppPlatform.cpp

```diff
@@ -0,0 +1,258 @@
+    {
+        if (NULL == mContentApps[index])
+        {
+            mContentApps[index] = app;
+            EmberAfStatus ret;
+            while (1)
+            {
+                ret = emberAfSetDynamicEndpoint(index, mCurrentEndpointId, ep, deviceType, DEVICE_VERSION_DEFAULT);
+                if (ret == EMBER_ZCL_STATUS_SUCCESS)
+                {
+                    ChipLogProgress(DeviceLayer, "Added device %s to dynamic endpoint %d (index=%d)",
+                                    app->GetApplicationBasic()->GetApplicationName(), mCurrentEndpointId, index);
+                    app->SetEndpointId(mCurrentEndpointId);
+                    return index;
+                }
+                else if (ret != EMBER_ZCL_STATUS_DUPLICATE_EXISTS)
+                {
+                    ChipLogProgress(DeviceLayer, "Adding device error=%d", ret);
+                    return -1;
+                }
+                // Handle wrap condition
+                if (++mCurrentEndpointId < mFirstDynamicEndpointId)
+                {
+                    mCurrentEndpointId = mFirstDynamicEndpointId;
+                }
+            }
+        }
+        index++;
+    }
+    ChipLogProgress(DeviceLayer, "Failed to add dynamic endpoint: No endpoints available!");
+    return -1;
+}
+
+int AppPlatform::RemoveContentApp(ContentApp * app)
+{
+    uint8_t index = 0;
+    while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
+    {
+        if (mContentApps[index] == app)
+        {
+            EndpointId ep       = emberAfClearDynamicEndpoint(index);
+            mContentApps[index] = NULL;
+            ChipLogProgress(DeviceLayer, "Removed device %s from dynamic endpoint %d (index=%d)",
+                            app->GetApplicationBasic()->GetApplicationName(), ep, index);
+            // Silence complaints about unused ep when progress logging
+            // disabled.
+            UNUSED_VAR(ep);
+            return index;
+        }
+        index++;
+    }
+    return -1;
+}
+
+void AppPlatform::SetupAppPlatform()
+{
+    ChipLogProgress(DeviceLayer, "AppPlatform::SetupAppPlatform()");
+
+    // Clear out the device database
+    uint8_t index = 0;
+    while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
+    {
+        mContentApps[index] = NULL;
+        index++;
+    }
+
+    // Set starting endpoint id where dynamic endpoints will be assigned, which
+    // will be the next consecutive endpoint id after the last fixed endpoint.
+    mFirstDynamicEndpointId = static_cast<chip::EndpointId>(
+        static_cast<int>(emberAfEndpointFromIndex(static_cast<uint16_t>(emberAfFixedEndpointCount() - 1))) + 1);
+    mCurrentEndpointId = mFirstDynamicEndpointId;
+
+    {
+        for (int i = 0; i < emberAfFixedEndpointCount(); i++)
+        {
+            ChipLogProgress(DeviceLayer, "endpoint index=%d, id=%d", i,
+                            static_cast<chip::EndpointId>(static_cast<int>(emberAfEndpointFromIndex(static_cast<uint16_t>(i)))));
```

Comment:
Why is this taking EndpointId (what emberAfEndpointFromIndex returns), casting it to int, then casting it back to EndpointId?

Also, why is this not using an `i` of type `uint16_t` to avoid needing _that_ cast?  And to not have `-Wconversion` problems.

Reply:
leftover debug method - removed in 12969